### PR TITLE
#4 Members clock diff

### PR DESCRIPTION
**Pull request by avkapustin**
#4 Add information about clock difference between members

### CODE REVIEW

File: membership.lua [hunk 1]
```diff
@@ -377,14 +381,27 @@ local function wait_ack(uri, ts, timeout)
 
         for _, ack in ipairs(_ack_cache) do
             if ack.dst == uri and ack.ts == ts then
-                return true
+                return true, ack
```
Comment (rosik) [inline membership.lua]: Boolean is redundant, use `return ack` and `return nil`

File: membership.lua [hunk 2]
```diff
@@ -377,14 +381,27 @@ local function wait_ack(uri, ts, timeout)
...
+    checks('table')
+    local ack_ts = tonumber(ack_data.ats)
+    local recv_ts = tonumber(ack_data.rts)
+    local start_ts = tonumber(ack_data.ts)
+
+    if ack_ts == nil or recv_ts == nil or start_ts == nil then
+        return nil
+    end
+
+    return math.abs(ack_ts - (recv_ts + start_ts) / 2)
```
Comment (rosik) [inline membership.lua]: No need for `abs`. Think of it the same way you think about time zones - GMT+3, GMT-6

File: membership.lua [hunk 3]
```diff
@@ -448,11 +469,17 @@ local function protocol_step()
...
+        local ok, ack_data = wait_ack(uri, loop_now, opts.PROTOCOL_PERIOD_SECONDS * 1.0e6)
+        if ok then
+            local member = members.get(uri)
+            -- calculate time difference between local time and member time
+            local delta = _get_clock_delta(ack_data)
+            members.set(uri, member.status, member.incarnation, { clock_delta = delta })
+            return
+        end
+    end
+    if members.get(uri).status == opts.ALIVE then
```
Comment (rosik) [inline membership.lua]: `elseif` was important. Imagine you received ack after indirect ping. Don't mark it suspect.

File: membership.lua [hunk 4]
```diff
@@ -662,6 +689,9 @@ end
 -- which corresponds to the last update of status or incarnation;
 -- it is always local and does not depend on other members’ clock setting.
 --
+-- @tfield number clock_delta difference of clocks (fiber.time64) between self and peer
+-- calculates while ping/ack protocol step or while probe_uri call
```
Comment (rosik) [inline membership.lua]: calculated during

File: membership/members.lua [hunk 5]
```diff
@@ -87,8 +88,9 @@ function members.set(uri, status, incarnation, payload)
     _all_members[uri] = {
         status = status,
         incarnation = incarnation,
-        payload = payload or (member or {}).payload,
+        payload = (params or {}).payload or (member or {}).payload,
```
Comment (rosik) [inline membership/members.lua]: Let's make this part more economical for the sake of gc: instead of creating empty tables write a couple of ifs afterwards

**Comment by knazarov** [comment]
@rosik Why anti-entrypy shouldn't send time delta? It looks fine.

**Comment by rosik** [comment]
Because it's local parameter - time difference between my clock and yours. An a third person has nothing to do with this information.

**Comment by knazarov** [comment]
We can then use the difference between our clock and the third person's clock to adjust the offsets in their table. This will help propagate time drifts faster and more reliably.

**Comment by knazarov** [comment]
Anyway, I suggest there should be a "generation" counter in time drifts anyway, because otherwise it's impossible to know in which direction to adjust offsets in our table.

**Comment by rosik** [comment]
Well, I think there is a way to to it. But can we proceed in a separate PR?

File: membership.lua (hunk 3 above)
Comment (avkapustin) [reply membership.lua]: If we receive indirect ACK, we return from protocol_step 2 lines above
But ok, I'll change this logic to get closer to original code

Comment (rosik) [reply membership.lua]: You're right

**Comment by avkapustin** [comment]
> `send_anti_entropy`
As I can understand, send_anti_entropy use members.pack to add member information, which is not affected by this PR

**Comment by rosik** [comment]
> > `send_anti_entropy`
> > As I can understand, send_anti_entropy use members.pack to add member information, which is not affected by this PR

You're right, I didn't notice that

**Pull request merged**